STORM_JSTORM_FEATURE_DIFF.md

```diff
@@ -0,0 +1,30 @@
+Monitor of task execute thread | N/A | Monitor the status of the execute thread of task. It is effective to find the slow bolt in a topology, and potential dead lock. | 	
+Message processing | Configurable multi receiving threads of worker (But likely to change to have deserialization happen in Netty) | <ol><li>Add receiving and transferring queue/thread for each task to make deserialization and serialization asynchronously<li>Remove receiving and transferring thread on worker level to avoid unnecessary locks and to shorten the message processing phase | Low risk for merging to JStorm (Both implementations are similar) |
+Batch tuples | Batching in DisruptorQueue | Do batch before sending tuple to transfer queue and support to adjust the batch size dynamically according to samples of actual batch size sent out for past intervals. | Should evaluate both implementations, and see which is better, or if we can pull in some of the dynamic batching to the disruptor level |	
+Grouping | Load aware balancing shuffle grouping | <ol><li>Add localfrist grouping that tuples are sent to the tasks in the same worker firstly. If the load of all local tasks is high, the tuples will be sent out to remote tasks.<li>Improve localOrShuffle grouping that tuples are only sent to the tasks in the same worker or in same node.	The shuffle of JStorm checks the load of network connection of local worker. | The risk for merging of  checking remote load which was implemented in "Load aware balancing shuffle grouping" is low.
+Rest API | YES | NA |The porting of rest API to JStorm is in progress
+web-UI | | <ol><li>Redesign Web-UI, the new Web-UI code is clear and clean.<li>Data display is much more beautiful and better user experience, especially topology graph and 30 minutes metric trend. url: http://storm.taobao.org/ 	
+metric system | | <ol><li>All levels of metrics, including stream metrics, task metrics, component metrics, topology metrics, even cluster metrics, will be sampled & calculated. Some metrics, e.g. ""tuple life cycle"", are very useful for debug and finding hot spot of a topology.<li>Support full metrics data. previous metric system can only display mean value of meters/histograms, the new metric system can display m1, m5, m15 of meters, and common percentiles of histograms.<li>Use new metrics windows, the min metric window is 1min, thus we can see the metircs data every single minute.<li>Supplies a metric uploader interface, thirty-party companies can easily build their own metric systems based on all historic metric data. | JStorm metric system is completely redesigned.
+jar command | Support progress bar when submitting topology | Does not support progress bar | Low risk for merging to JStorm
+rebalance	command | Basic functionality of rebalance | Besides rebalance, scale-out/in by updating the number of worker, acker, spout & bolt dynamically without stopping topology | How is grouping consistency maintined when changing the number or bolt/spout? | 

```

Comment:
There are no plans to change exactly once processing right now.  At yahoo we have a decent number of trident topologies running in production, and I don't see it ever going fully away.  But I do see the appeal of having something like apex or flink, where you can have exactly once processing without the requirement for kafka to do the replay.  It would also be nice to be able to have the exact same code work for both exactly once, at least once, and best effort processing.

But that would be a long ways off if we do it.